### DISCUSSION TITLE
Trying to display LHCb event data

### DISCUSSION THREAD

**Discussion by andrewpap22** [HSF/phoenix#274] [Q&A] [answered]
### Greetings, 

So, as discussed on the Wednesday meeting, I've been trying to display some LHCb event data that I've gathered from C++ code of the detector itself. ( of course with the help of my supervisor @sponce )

I think that the format of the file is correct ( but of course I might be wrong since nothing is being displayed ). 

`here is my file (uploaded in .txt because GitHub doesn't support upload of .json in discussion)` [LHCb_event_data.txt](https://github.com/HSF/phoenix/files/6325075/LHCb_event_data.txt)

The ways I tried to display my file are the following: 

 - Load the file from the GUI interface directly
 - Change the default LHCb file with mine in the corresponding typescript file. 

The part of the code where I changed the file is the following: 

```javascript
/**
 * Directory Path: 
 * /phoenix/packages/phoenix-ng/projects/phoenix-app/src/assets/files/lhcb
 * So this is the path where I added my file, 
 * And then on the path: 
 * /phoenix/packages/phoenix-ng/projects/phoenix-app/src/app/sections/lhcb
 * The modified file: lhcb.component.ts
*/

// Line 60
.get("assets/files/lhcb/my_file.json")
```

Any help with what I'm doing wrong would be appreciated, and I'll be glad to help and contribute to the phoenix project as I progress with my knowledge and all... 

*Thank you in advance for your valuable help,*

### All the best, 

Andreas

**Comment by EdwardMoyse** [comment]
Hi @andrewpap22 - firstly, you cannot load this file from the GUI directly because it's not in Phoenix format.

As an example of a file which is, have a look at :
![image](https://user-images.githubusercontent.com/6764617/115387489-0e043a80-a1db-11eb-8e77-60414dc9f012.png)

which is this one:
https://github.com/HSF/phoenix/blob/master/packages/phoenix-ng/projects/phoenix-app/src/assets/files/event_data/ATLAS_calibration.json

We should add a LHCb loader to the GUI. #281

I'll try your second approach now.

**Reply by EdwardMoyse** [reply]
BTW we should improve reporting of errors in Phoenix, but if you open your console log in your browser you might have seen something. And if you open the info panel, you can get a hint that something is wrong because the event number and run number are undefined.

![image](https://user-images.githubusercontent.com/6764617/115389135-03e33b80-a1dd-11eb-975f-d0bf61f69248.png)

But we should definitely try to make this clearer.

**Reply by EdwardMoyse** [reply]
Incidentally, the earlier "Error: Invalid file format!" in the info panel is because the file was named `.json.txt` when I first tried to open it (I'm using Windows and I still struggle with the basics!)

**Comment by EdwardMoyse** [comment]
Okay, this file also seems to be a completely different json format from the file provided by @bcouturi, which is what the LHCb loader understands.

Here's the file Ben provided:
https://github.com/HSF/phoenix/blob/master/packages/phoenix-ng/projects/phoenix-app/src/assets/files/lhcb/00191749_0005296728.json

How did you generate this JSON file? Perhaps @bcouturi can help?

Otherwise, if you want to look at the ATLAS Phoenix dumper it's here:
https://gitlab.cern.ch/atlas/athena/-/tree/master/Event/DumpEventDataToJSON/src

It's obviously reading in ATLAS data, but I think it shouldn't be too hard to port it to LHCb.

**Comment by EdwardMoyse** [comment]
I'm very happy to have a chat over zoom or whatever, if this doesn't make sense!

**Comment by bcouturi** [comment]
Sorry just catching up as I was off yesterday. Andreas is redoing (cleanly) the whole chain so the idea is to abandon our custom export/import for something more standard if possible.

**Comment by andrewpap22** [comment, marked as answer]
![pic-selected-210429-1550-49](https://user-images.githubusercontent.com/68736032/116533337-ec493880-a8e9-11eb-91a0-395aa35086b9.png)

### Greetings, 

So I managed to make the .json file proper in Phoenix format (of course with the valuable help of Dr. @sponce) and as you can see in the screenshot now we have a proper visualization for Long tracks and VPHits for the LHCb experiment as well! 😄 

Now for our next steps I would like again to ask for your valuable help in trying to properly display as well the `Calorimeters` and `Energy Deposits`. 

As you can see in our new created file below, (`remove the .txt extention when you download the file`)

[tracksV2.json.txt](https://github.com/HSF/phoenix/files/6397745/tracksV2.json.txt)

We have inside that file our Calorimeters (ECal and HCal) and they're not like the ATLAS ones where you have eta and phi, etc... we instead have XYZ points and of course the energy deposits as you have. We could convert the XYZ points to phi I guess but that would not make sense for the LHCb Clusters. 

So a workaround would be I guess to write some code in the Phoenix library in the LHCb part of the project to understand and display properly our LHCb specific CaloClusters. 

So my question is if you could "drive" me somehow on trying to start and write that particular code (I would be grateful if you could give me some directions to try and do that inside Phoenix, because it will be a great learning journey for me and obviously far than interesting) or if there is something else we can do here and of course again I want to invest my time in doing that! 😄 

Thank you very much for your time and help!!!

**Reply by 9inpachi** [reply]
Hi @andrewpap22,

Great to hear that the tracks are working now! Congrats! ;)

For handling Calorimeters differently, you could extend the `LHCbLoader` to have a "new" physics object handled entirely by the `LHCbLoader`. This is something I have already done with CMS experiment for the `MuonChambers` so I am sure it will be a really good reference for you.

Fortunately, I also created a dedicated guide which you can read through to understand how adding new physics objects work. The guide is "[Handling new physics objects](https://github.com/HSF/phoenix/blob/master/guides/developers/event-data-loader.md#handling-new-physics-objects)". I am pretty sure it's the perfect solution for your use case.
But if you still have questions, feel free to throw them -  we would love to help.

For a simple explanation on how I did it for the CMS experiment. You can take a look at the [`loadObjectTypes`](https://github.com/HSF/phoenix/blob/master/packages/phoenix-event-display/src/loaders/cms-loader.ts#L28) method of [`CMSLoader`](https://github.com/HSF/phoenix/blob/master/packages/phoenix-event-display/src/loaders/cms-loader.ts) which is a method overridden from the parent `PhoenixLoader` class. This method calls the parent method and loads all object types supported by the parent and immediately after, loads the `MuonChambers` that are specific to `CMSLoader`.

Good luck on implementing that! And again, feel free to ask any questions. ;)

**Reply by andrewpap22** [reply]
Hello there @9inpachi, 

Thank you very much for your valuable help and instructions, I will try to implement it and once done I will make a PR and update you with any news or questions I might have! 

Again thank you!! 😄

**Reply by EdwardMoyse** [reply]
I agree with this plan - but I wonder if we can make it general and not LHCb specific (of course we can also do this afterwards!)

We could be to introduce a new phoenix 'physics object' `PlanarCaloCell`, where you also specify a plane in the data i.e. something like

```
PlanarCaloCell : [
   "HCalo": {
      "Plane" : [x,y,z,d].
      "HCal_digit": [
...
```
Where the plane would be defined with usual threejs notation:
https://threejs.org/docs/index.html?q=Plane#api/en/math/Plane

The only thing if we did this is I would prefer that the notation was more Phoenixlike ... i.e. pos instead of `XYZPoint`. We should discuss the format and agree on it in this case.

We could also rename the existing CaloCluster to CylindricalCaloCell and similarly introduce a cylindrical bound to draw the cells (as was initially planned) i.e. 

```
PlanarCaloCell : [
   "HCalo": {
      "Cylinder" : [x,y,z,l,r]. // centre and length, radius
      "HCal_digit": [
...
```

**Reply by sponce** [reply]
We should definitely do something reusable if feasible. LHCb is not the only experiment with planar calorimeters for sure !
Now in your proposal, what would be the content of `Hcal_digit` ? Something like {"pos" :[x,y], "energy":e, "otherdecoration":"whatever"} ?

**Reply by sponce** [reply]
And [x,y] would be [eta,phi] in the Cylinder case of course

**Reply by EdwardMoyse** [reply]
Hi @sponce - That looks good for the content of the cell. So maybe something like this?
```
PlanarCaloCells : [
   "HCalo": {
      "plane" : [x,y,z,d].
      "cells": [ {"pos" :[x,y], "energy":e, "otherdecoration":"whatever"}, {}, {} ]
     },
    "ECalo": {
      "plane" : [x,y,z,d].
      "cells": [ {"pos" :[x,y], "energy":e, "otherdecoration":"whatever"}, {}, {} ]
     },
]
```
i.e. you have the object type, then the names of the containers, with the plane coords and an attribute cells, which contains the cell objects

If they're all the same size, we could make that a property of the collection i.e.
```
   "HCalo": {
      "plane" : [x,y,z,d].
      "size": [0.1,0.1],
      "cells": [ {"pos" :[x,y], "energy":e, "otherdecoration":"whatever"}, {}, {} ]
     },
```
If they vary, then it could be a property of the cell itself.

**Reply by sponce** [reply]
Looks good to me. The `size` attribute at the top level could be seen as a default size which can be overwritten by a specific one in a given cell. @andrewpap22 do you feel like trying to implement that ? I obviously will help !

**Reply by 9inpachi** [reply]
If LHCb is not the only experiment with `PlanarCaloCell` then we should definitely have it in the `PhoenixLoader`.

For a quick direction, the code to draw the 3D object will go to [`PhoenixObjects`](https://github.com/HSF/phoenix/blob/master/packages/phoenix-event-display/src/loaders/objects/phoenix-objects.ts) and then we can load the new object by extending `loadObjectTypes` to include `PlanarCaloCell` [like we do for tracks](https://github.com/HSF/phoenix/blob/master/packages/phoenix-event-display/src/loaders/phoenix-loader.ts#L147-L165).

**Reply by andrewpap22** [reply]
Alright then I updated my notes as well and I'll be trying to implement what was discussed here! I will keep you updated! 😃
